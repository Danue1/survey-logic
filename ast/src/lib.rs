mod comparison;
mod condition;
mod literal;
mod program;
mod statement;
mod visitor;

pub use comparison::*;
pub use condition::*;
pub use literal::*;
pub use program::*;
pub use statement::*;
pub use visitor::*;
