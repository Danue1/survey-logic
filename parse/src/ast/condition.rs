use crate::context::{Context, State};
use syntax_kind::SyntaxKind;

impl Context {
    pub fn logical_condition(&mut self) -> State {
        let checkpoint = self.checkpoint();
        if self.atomic_condition() == State::Continue {
            return State::Continue;
        }

        while !self.is_empty() {
            self.skip_whitespace();
            if self.logical_operator() == State::Continue {
                break;
            }

            self.skip_whitespace();
            self.start_node_at(checkpoint, SyntaxKind::LOGICAL_CONDITION_NODE);
            self.atomic_condition();
            self.finish_node();
        }

        State::Stop
    }

    pub fn atomic_condition(&mut self) -> State {
        one_of!(self.group_condition(), self.binary_condition(),)
    }

    pub fn group_condition(&mut self) -> State {
        if self.peek() != SyntaxKind::LEFT_PAREN {
            return State::Continue;
        }

        self.start_node(SyntaxKind::GROUP_CONDITION_NODE);
        self.bump();
        self.skip_whitespace();
        if self.logical_condition() != State::Continue {
            self.skip_whitespace();
        }
        if self.peek() == SyntaxKind::RIGHT_PAREN {
            self.bump();
        }

        self.finish_node();
        State::Stop
    }

    pub fn logical_operator(&mut self) -> State {
        match self.peek_str().to_lowercase().as_str() {
            "and" => {
                self.start_node(SyntaxKind::AND_OPERATOR_NODE);
                self.bump();
                self.finish_node();
                State::Stop
            }
            "or" => {
                self.start_node(SyntaxKind::OR_OPERATOR_NODE);
                self.bump();
                self.finish_node();
                State::Stop
            }
            _ => State::Continue,
        }
    }

    pub fn binary_condition(&mut self) -> State {
        let checkpoint = self.checkpoint();
        if self.question_parameter() == State::Continue {
            return State::Continue;
        }

        self.start_node_at(checkpoint, SyntaxKind::COMPARISON_CONDITION_NODE);
        self.skip_whitespace();

        if let Some(kind) = self.punctuation() {
            self.skip_whitespace();
            match kind {
                PunctuationKind::Comparison => {
                    self.start_node_at(checkpoint, SyntaxKind::BINARY_COMPARISON_NODE);
                    self.singular_literal();
                    self.finish_node();
                }
                PunctuationKind::In => {
                    self.start_node_at(checkpoint, SyntaxKind::LIST_COMPARISON_NODE);
                    self.list_literal();
                    self.finish_node();
                }
                PunctuationKind::Date => {
                    self.start_node_at(checkpoint, SyntaxKind::DATE_COMPARISON_NODE);
                    self.date_literal();
                    self.finish_node();
                }
            };
        }

        self.finish_node();
        State::Stop
    }

    pub fn punctuation(&mut self) -> Option<PunctuationKind> {
        match self.peek_str().to_lowercase().as_str() {
            "=" => {
                self.start_node(SyntaxKind::EQUAL_OPERATOR_NODE);
                self.bump();
                self.finish_node();
                return Some(PunctuationKind::Comparison);
            }
            "!" => {
                let checkpoint = self.checkpoint();
                self.bump();
                if self.peek_str() == "=" {
                    self.start_node_at(checkpoint, SyntaxKind::NOT_EQUAL_OPERATOR_NODE);
                    self.bump();
                    self.finish_node();
                    return Some(PunctuationKind::Comparison);
                }

                self.start_node_at(checkpoint, SyntaxKind::ERROR);
                self.finish_node();
                return None;
            }
            comparison @ ("<" | ">") => {
                let checkpoint = self.checkpoint();
                self.bump();
                if self.peek_str() == "=" {
                    if comparison == "<" {
                        self.start_node_at(checkpoint, SyntaxKind::LESS_THAN_OR_EQUAL_OPERATOR_NODE)
                    } else {
                        self.start_node_at(
                            checkpoint,
                            SyntaxKind::GREATER_THAN_OR_EQUAL_OPERATOR_NODE,
                        )
                    };
                    self.bump();
                    self.finish_node();
                    return Some(PunctuationKind::Comparison);
                }

                if comparison == "<" {
                    self.start_node_at(checkpoint, SyntaxKind::LESS_THAN_OPERATOR_NODE);
                } else {
                    self.start_node_at(checkpoint, SyntaxKind::GREATER_THAN_OPERATOR_NODE);
                };
                self.finish_node();
                return Some(PunctuationKind::Comparison);
            }
            "in" => {
                self.start_node(SyntaxKind::IN_OPERATOR_NODE);
                self.bump();
                self.finish_node();
                return Some(PunctuationKind::In);
            }
            "after" => {
                self.start_node(SyntaxKind::AFTER_OPERATOR_NODE);
                self.bump();
                self.finish_node();
                return Some(PunctuationKind::Date);
            }
            "before" => {
                self.start_node(SyntaxKind::BEFORE_OPERATOR_NODE);
                self.bump();
                self.finish_node();
                return Some(PunctuationKind::Date);
            }
            _ => None,
        }
    }
}

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum PunctuationKind {
    Comparison,
    In,
    Date,
}

#[cfg(test)]
mod tests {
    macro_rules! assert_snapshot {
        ($parse:ident($source:expr)) => {
            let tokens = lex::lex($source);
            let mut context = super::Context::new(tokens);
            context.$parse();
            let node = context.finish();
            insta::assert_debug_snapshot!(node);
        };
    }

    #[test]
    fn comparison() {
        assert_snapshot!(binary_condition("Q123 = 1"));
        assert_snapshot!(binary_condition("Q123 != 1"));
        assert_snapshot!(binary_condition("Q123 < 1"));
        assert_snapshot!(binary_condition("Q123 > 1"));
        assert_snapshot!(binary_condition("Q123 <= 1"));
        assert_snapshot!(binary_condition("Q123 >= 1"));
    }

    #[test]
    fn group() {
        assert_snapshot!(group_condition("(Q123 = 1)"));
        assert_snapshot!(group_condition("((Q123 = 1))"));
    }

    #[test]
    fn logical() {
        assert_snapshot!(logical_condition("Q123 = 1 and Q456 = 2"));
        assert_snapshot!(logical_condition("Q123 = 1 and Q456 = 2 or Q789 = 3"));
    }

    #[test]
    fn complex() {
        assert_snapshot!(logical_condition("(Q123 = 1) and Q456 = 2"));
        assert_snapshot!(logical_condition("Q123 = 1 and (Q456 = 2)"));
        assert_snapshot!(logical_condition("(Q123 = 1) and (Q456 = 2)"));
        assert_snapshot!(logical_condition("(Q123 = 1 and Q456 = 2)"));
        assert_snapshot!(logical_condition("(Q123 = 1 and Q456 = 2) or Q789 = 3"));
        assert_snapshot!(logical_condition("Q123 = 1 and (Q456 = 2 or Q789 = 3)"));
    }
}
