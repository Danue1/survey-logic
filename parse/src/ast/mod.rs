mod condition;
mod literal;
mod question_parameter;
mod statement;

use crate::context::{Context, State};
use syntax_kind::SyntaxKind;

impl Context {
    pub fn program(&mut self) {
        self.start_node(SyntaxKind::PROGRAM_NODE);
        self.skip_whitespace();
        while !self.is_empty() {
            if self.statement() == State::Continue {
                self.unexpected_token();
            }
            self.skip_whitespace();
        }
        self.finish_node();
    }

    pub fn unexpected_token(&mut self) {
        self.start_node(SyntaxKind::ERROR);
        self.bump();
        self.finish_node();
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn statements() {
        let tokens = lex::lex("Q123 = 1 Q456 = 2");
        let mut context = super::Context::new(tokens);
        context.program();
        let node = context.finish();
        insta::assert_debug_snapshot!(node);
    }
}
