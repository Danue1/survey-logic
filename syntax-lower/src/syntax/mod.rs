mod comparison;
mod condition;
mod literal;
mod program;
mod statement;

pub use comparison::*;
pub use condition::*;
pub use literal::*;
pub use program::*;
pub use statement::*;
