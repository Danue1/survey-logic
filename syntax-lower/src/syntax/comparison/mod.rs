mod binary_comparison;
mod date_comparison;
mod list_comparison;

use crate::SyntaxResult;
use ast::{Comparison, ComparisonKind};
pub use binary_comparison::*;
pub use date_comparison::*;
pub use list_comparison::*;
use syntax::{ComparisonKindNode, ComparisonNode};

pub fn lower_comparison(node: ComparisonNode) -> SyntaxResult<Comparison> {
    let kind = match node.kind() {
        Some(kind) => lower_comparison_kind(kind)?,
        None => invalidate!("Expected a comparison kind"),
    };
    Ok(Comparison::new(node.syntax(), node.parent(), kind))
}

pub fn lower_comparison_kind(node: ComparisonKindNode) -> SyntaxResult<ComparisonKind> {
    match node {
        ComparisonKindNode::Binary(node) => {
            let comparison = crate::lower_binary_comparison(node)?;
            Ok(ComparisonKind::Binary(comparison))
        }
        ComparisonKindNode::List(node) => {
            let comparison = crate::lower_list_comparison(node)?;
            Ok(ComparisonKind::List(comparison))
        }
        ComparisonKindNode::Date(node) => {
            let comparison = crate::lower_date_comparison(node)?;
            Ok(ComparisonKind::Date(comparison))
        }
    }
}
