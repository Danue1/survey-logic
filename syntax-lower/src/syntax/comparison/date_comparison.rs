use crate::SyntaxResult;
use ast::DateComparison;
use syntax::DateComparisonNode;

pub fn lower_date_comparison(node: DateComparisonNode) -> SyntaxResult<DateComparison> {
    let parameter = match node.parameter() {
        Some(parameter) => crate::lower_question_parameter(parameter)?,
        None => invalidate!("Expected a parameter"),
    };
    let value = match node.value() {
        Some(value) => crate::lower_date_literal(value)?,
        None => invalidate!("Expected a value"),
    };
    Ok(DateComparison::new(
        node.syntax(),
        node.parent(),
        parameter,
        value,
    ))
}

#[cfg(test)]
mod tests {
    macro_rules! assert_snapshot {
        ($source:expr) => {
            let tokens = lex::lex($source);
            let node = parse::parse(tokens);
            let ast = crate::lower(node);
            insta::assert_debug_snapshot!(ast);
        };
    }

    #[test]
    fn before() {
        assert_snapshot!("Q123 BEFORE \"2023-11-15\"");
    }

    #[test]
    fn after() {
        assert_snapshot!("Q123 AFTER \"2023-11-15\"");
    }
}
