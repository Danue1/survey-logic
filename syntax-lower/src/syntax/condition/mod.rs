mod comparison_condition;
mod group_condition;
mod logical_condition;

use crate::SyntaxResult;
use ast::{Condition, ConditionKind};
pub use comparison_condition::*;
pub use group_condition::*;
pub use logical_condition::*;
use syntax::{ConditionKindNode, ConditionNode};

pub fn lower_condition(node: ConditionNode) -> SyntaxResult<Condition> {
    let kind = match node.kind() {
        Some(kind) => lower_condition_kind(kind)?,
        None => invalidate!("Expected a condition kind"),
    };
    Ok(Condition::new(node.syntax(), node.parent(), kind))
}

pub fn lower_condition_kind(node: ConditionKindNode) -> SyntaxResult<ConditionKind> {
    match node {
        ConditionKindNode::Group(group) => {
            let group = crate::lower_group_condition(group)?;
            Ok(ConditionKind::Group(Box::new(group)))
        }
        ConditionKindNode::Logical(logical) => {
            let logical = crate::lower_logical_condition(logical)?;
            Ok(ConditionKind::Logical(Box::new(logical)))
        }
        ConditionKindNode::Comparison(comparison) => {
            let comparison = crate::lower_comparison_condition(comparison)?;
            Ok(ConditionKind::Binary(Box::new(comparison)))
        }
    }
}
