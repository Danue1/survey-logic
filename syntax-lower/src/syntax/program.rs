use crate::SyntaxResult;
use ast::Program;
use syntax::ProgramNode;

pub fn lower_program(node: ProgramNode) -> SyntaxResult<Program> {
    let mut statements = vec![];
    for statement in node.statements() {
        statements.push(crate::lower_statement(statement)?);
    }
    Ok(Program::new(node.syntax(), statements))
}
