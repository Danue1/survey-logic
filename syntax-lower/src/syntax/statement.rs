use crate::SyntaxResult;
use ast::Statement;
use syntax::StatementNode;

pub fn lower_statement(node: StatementNode) -> SyntaxResult<Statement> {
    let condition = match node.condition() {
        Some(condition) => crate::lower_condition(condition)?,
        None => invalidate!("Expected a condition"),
    };
    Ok(Statement::new(node.syntax(), node.parent(), condition))
}
