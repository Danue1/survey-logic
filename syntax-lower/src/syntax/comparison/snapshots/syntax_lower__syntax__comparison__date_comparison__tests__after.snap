---
source: syntax-lower/src/syntax/comparison/date_comparison.rs
expression: ast
---
Ok(
    Program {
        statements: [
            Statement {
                condition: Condition {
                    kind: Binary(
                        ComparisonCondition {
                            comparison: Comparison {
                                kind: Date(
                                    DateComparison {
                                        parameter: QuestionParameter {
                                            id: NumericLiteral {
                                                value: 123,
                                            },
                                        },
                                        value: DateLiteral {
                                            value: "\"2023-11-15\"",
                                        },
                                    },
                                ),
                            },
                        },
                    ),
                },
            },
        ],
    },
)
